### PR TITLE
[platform][app-arch] - update CRD to support multiple jobs per class

### PR DESCRIPTION
**Pull request by hlokavarapu**
**Why is this change needed?**
Although WPA-multiqueue supports scaling based on multiple SQS queues, teams are enqueing different messages within the same queue.

**How does the PR address the problem?**
Update the CRD spec to support configuration per job class and not per queue. 

**What else do I need to know?**
https://k2labs.atlassian.net/browse/MIG-4699

### CODE REVIEW

File: artifacts/examples/shoryuken-worker-wpa-multiqueue.yaml [hunk 1]
```diff
@@ -0,0 +1,20 @@
+apiVersion: k8s.practo.dev/v1
+kind: WorkerPodAutoScalerMultiQueue
+metadata:
+  name: shoryuken-worker-wpa-multiqueue-development
+spec:
+  minReplicas: 1
+  maxReplicas: 10
+  deploymentName: shoryuken-worker-development
+  queues:
```
Comment (hlokavarapu) [inline artifacts/examples/shoryuken-worker-wpa-multiqueue.yaml:9]: Here is an example of the new spec for wpa-multiqueue. I have introduced a new variable called `enqueueRate` per job. This is then used internally to calculate `targetMessagesPerWorker`.

File: pkg/queue/queue.go [hunk 2]
```diff
@@ -177,6 +178,15 @@ func (q *Queues) Add(namespace string, name string, uri string,
 		idleWorkers = spec.idleWorkers
 	}
 
+	approxSecondsToProcessOneJob := float64(0)
+	totalEnqueueRate := float64(0)
+	for _, job := range jobs {
```
Comment (hlokavarapu) [inline pkg/queue/queue.go:183]: Calculate seconds to process a job based on provided enqueue rates per job. Job enqueue rate must be greater than 0.1.

File: pkg/controller/controller.go [hunk 3]
```diff
@@ -625,11 +625,13 @@ func getMinMultiQueueWorkers(
...
 	var totalMinWorkers int32
+	var totalWorkersBasedOnMessagesSent float64
 	for _, qSpec := range queueSpecs {
-		workersBasedOnMessagesSent := int32(math.Ceil((qSpec.SecondsToProcessOneJob * qSpec.MessagesSentPerMinute) / 60))
+		workersBasedOnMessagesSent := (qSpec.SecondsToProcessOneJob * qSpec.MessagesSentPerMinute) / 60
+		totalWorkersBasedOnMessagesSent += workersBasedOnMessagesSent
 		klog.V(4).Infof("%s: secondsToProcessOneJob=%v, workersBasedOnMessagesSent=%v\n", qSpec.Name, qSpec.SecondsToProcessOneJob, workersBasedOnMessagesSent)
-		totalMinWorkers += workersBasedOnMessagesSent
 	}
+	totalMinWorkers = int32(math.Ceil(totalWorkersBasedOnMessagesSent))
```
Comment (hlokavarapu) [inline pkg/controller/controller.go:657]: improve accuracy of total min worker calculation

File: pkg/controller/controller.go [hunk 4]
```diff
@@ -790,13 +793,14 @@ func GetDesiredWorkersMultiQueue(
...
 	var desiredWorkers int32
 	for _, k8QSpec := range k8QueueSpecs {
 		qSpec := queueSpecs[k8QSpec.URI]
-		desiredWorkers += int32(math.Ceil(
-			float64(qSpec.Messages) / float64(k8QSpec.TargetMessagesPerWorker)),
-		)
+		targetMessagesPerWorker := float64(k8QSpec.SLA) / qSpec.SecondsToProcessOneJob
+		approxDesiredWorkers += float64(qSpec.Messages) / targetMessagesPerWorker
 	}
+	desiredWorkers = int32(math.Ceil(approxDesiredWorkers))
```
Comment (hlokavarapu) [inline pkg/controller/controller.go:836]: improve accuracy of desired worker calculation

File: artifacts/crd.yaml [hunk 5]
```diff
@@ -60,15 +60,45 @@ spec:
...
+                        type: object
+                        properties:
+                          className:
+                            type: string
+                            description: 'Job class name'
+                          secondsToProcessOneJob:
+                            type: number
+                            format: float
+                            description: 'For fast running workers doing high RPM, the backlog is very close to zero. So for such workers scale up cannot happen based on the backlog, hence this is a really important specification to always keep the minimum number of workers running based on the job RPM.'
+                          enqueueRate:
```
Comment (jibin-k-jose) [inline artifacts/crd.yaml]: Can we call this `peakEnqueueRate` instead ?

Comment (hlokavarapu) [reply artifacts/crd.yaml]: Refactored variable name

**Pull request merged**